File: backend/utils/distance.go
package utils

import "math"

func DistanceMeters(lat1, lon1, lat2, lon2 float64) float64 {
	rad := math.Pi / 180
	a := math.Pow(math.Sin((lat2-lat1)*rad/2), 2) + math.Cos(lat1*rad)*math.Cos(lat2*rad)*math.Pow(math.Sin((lon2-lon1)*rad/2), 2)
	a = math.Max(0, math.Min(1, a))
	return 6371000 * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

File: frontend/src/utils/distance.ts
import { Position } from '../store/useStore'
export function distance(a: Position, b: Position) {
  const rad = Math.PI/180
  const x = Math.sin((b.lat-a.lat)*rad/2)**2 + Math.cos(a.lat*rad)*Math.cos(b.lat*rad)*Math.sin((b.lon-a.lon)*rad/2)**2
  return 6371000*2*Math.atan2(Math.sqrt(x), Math.sqrt(Math.max(0,1-x)))
}
